src/libtorch.cc

```diff
@@ -929,6 +907,15 @@ ModelInstanceState::ReadOutputTensors(
         (device_.type() == torch::kCPU) ? TRITONSERVER_MEMORY_CPU
                                         : TRITONSERVER_MEMORY_GPU,
         (device_.type() == torch::kCPU) ? 0 : device_.index());
+
```

Comment:
You should be able to move this to line 897 above just after output_buffer pointer is assigned.

Reply:
I did try that first. I shared number for the same in the ticket. It appears the synchronization occurs during the responder.ProcessTensor call.

Reply:
Why would that be the case? Are you saying that the content in the output_buffer is not valid until within the responder? But the responder doesn't interact with the torch library at all, how does the synchronization happen?

Reply:
I am not 100% clear but this is what I believe is happening:
The output_buffer is a pointer to memory on GPU that the output tensor from pytorch will be stored. While trying to access data at the location, pytorch synchronizes and populates the output_buffer. While this is odd it does help with performance.
I could not find any documentation from pytorch that explains this behavior.

Reply:
There's no way for pytorch to do that. Once we have a char* pointer to the data, accessing it via that pointer will not trigger anything that pytorch can be aware of (if there was some OS level page protection on the memory there could be some OS action triggered... but not in the pytorch code).

Reply:
I can run some more tests to understand what causes this behavior.